File: front/src/routes/engine.ts
import { CellColor, GameStyle, type GameField, MoveResult } from './types';
import { TurnError } from './types';

function tossTwoDice() {
	const dice1 = Math.floor(Math.random() * 6) + 1;
	const dice2 = Math.floor(Math.random() * 6) + 1;

	return [dice1, dice2] as [number, number];
}

export function tossDiceInTurn(gameField: GameField): [TurnError, null] | [null, [number, number]] {
	if (gameField.tossed && gameField.tossed.length !== 0) {
		return [TurnError.DoubleTossIsNotAllowed, null];
	}
	const [dice1, dice2] = tossTwoDice();

	if (dice1 === dice2) {
		gameField.tossed = [dice1, dice1, dice1, dice1];
	} else {
		gameField.tossed = [dice1, dice2];
	}

	return [null, [dice1, dice2]];
}

export function initGameField() {
	const gameField: GameField = {
		cells: Array(24)
			.fill(0)
			.map(() => ({
				color: CellColor.Unoccupied,
				count: 0
			})),
		gameStyle: GameStyle.Long,
		turn: CellColor.White
	};

	gameField.cells[11].color = CellColor.White;
	gameField.cells[11].count = 15;

	gameField.cells[23].color = CellColor.Black;
	gameField.cells[23].count = 15;
	return gameField;
}

export function moveChecker(
	gameField: GameField,
	fromCellIndex: number,
	toCellIndex: number,
	colorToMove: CellColor
) {
	if (fromCellIndex > 23 || fromCellIndex < 0) {
		console.warn('Source cell index is out of bounds');
		return TurnError.SourceCellIsOutOfBoard;
	}

	if (toCellIndex > 23 || toCellIndex < 0) {
		console.warn('Target cell index is out of bounds');
		return TurnError.TargetCellIsOutOfBoard;
	}

	const fromCell = gameField.cells[fromCellIndex];

	if (fromCell.color !== colorToMove) {
		console.warn('Source cell is occupied by opponent or empty');
		console.log('colorToMove', colorToMove);
		console.log('fromCellIndex', fromCellIndex);
		console.log('fromCell', fromCell);

		if (fromCell.color === CellColor.Unoccupied) {
			return TurnError.SourceCellIsEmpty;
		} else {
			return TurnError.SourceCellIsOccupiedByOpponent;
		}
	}

	if (fromCell.count === 0) {
		console.warn('No checkers to move');
		return TurnError.SourceCellIsEmpty;
	}

	const toCell = gameField.cells[toCellIndex];

	if (toCell.color !== CellColor.Unoccupied && toCell.color !== colorToMove) {
		console.warn('Target cell is occupied by opponent');
		return TurnError.TargetCellIsOccupiedByOpponent;
	}

	fromCell.count--;
	toCell.count++;

	toCell.color = fromCell.color;

	if (fromCell.count === 0) {
		fromCell.color = CellColor.Unoccupied;
	}

	return;
}

export function moveCheckerBy(
	gameField: GameField,
	fromCellIndex: number,
	step: number,
	colorToMove: CellColor
) {
	let toCellIndex;

	toCellIndex = fromCellIndex - step;

	if (toCellIndex < 0) {
		toCellIndex = 24 + toCellIndex;
	}

	return moveChecker(gameField, fromCellIndex, toCellIndex, colorToMove);
}

export function playerMoveCheckerBy(
	gameField: GameField,
	fromCellIndex: number,
	step: number,
	colorToMove: CellColor
) {
	let toCellIndex = fromCellIndex - step;
	if (toCellIndex < 0) {
		toCellIndex = 24 + toCellIndex;
	}

	return playerMoveChecker(gameField, fromCellIndex, toCellIndex, colorToMove);
}

export function playerMoveChecker(
	gameField: GameField,
	fromCellIndex: number,
	toCellIndex: number,
	colorToMove: CellColor
): [TurnError, null] | [null, MoveResult] {
	const turn = gameField.turn;

	if (colorToMove !== turn) {
		console.warn('Wrong turn');
		return [TurnError.WrongTurn, null];
	}

	let step = fromCellIndex - toCellIndex;
	if (step < 0) {
		step = 24 + step;
	}

	if (!gameField.tossed?.includes(step)) {
		console.warn(`Step is not in tossed. Step: ${step}, tossed: ${gameField.tossed?.join(', ')}`);
		return [TurnError.InvalidPlayerMove, null];
	}

	const error = moveChecker(gameField, fromCellIndex, toCellIndex, colorToMove);

	if (error) {
		return [error, null];
	}

	const index = gameField.tossed.indexOf(step);
	gameField.tossed.splice(index, 1);

	if (gameField.tossed.length === 0) {
		gameField.turn = gameField.turn === CellColor.White ? CellColor.Black : CellColor.White;
		return [null, MoveResult.TurnIsOver];
	}

	return [null, MoveResult.HasNextMove];
}

File: backend/src/engine.go
package src

import (
	"fmt"
	"math/rand"
)

type GameField struct {
	// array of 24 cells
	cells   [24]Cell
	variant GameVariant
}

type Color int

const (
	None Color = iota
	Black
	White
)

type GameVariant string

const (
	Long GameVariant = "long"
)

type Cell struct {
	color    Color
	checkers int
}

func initField() GameField {
	cells := [24]Cell{}

	cells[11].checkers = 15
	// cells[10].checkers = 1
	cells[11].color = White
	// cells[10].color = White

	cells[23].checkers = 15
	cells[23].color = Black

	return GameField{cells, Long}
}

func (field *GameField) printField() {
	fmt.Println("Field:")
	for i := 0; i < 15; i++ {
		for j := 0; j < 12; j++ {
			topRow := field.cells[j]
			bottomRow := field.cells[23-j]
			if i < topRow.checkers {
				if topRow.color == White {
					fmt.Printf("⬜️")
				} else if topRow.color == Black {
					fmt.Printf("⬛️")
				} else {
					fmt.Printf("🟥")
				}
			} else if 14-i < bottomRow.checkers {
				if bottomRow.color == White {
					fmt.Printf("⬜️")
				} else if bottomRow.color == Black {
					fmt.Printf("⬛️")
				} else {
					fmt.Printf("🟥")
				}
			} else {
				fmt.Printf("🟫")
			}
		}
		fmt.Println()
	}
}

func tossDice() int {
	return rand.Intn(6) + 1
}

func tossTwoDice() (int, int) {
	return tossDice(), tossDice()
}

// move checher

func (field *GameField) moveChecker(from int, to int) {
	if from < 0 || from > 23 || to < 0 || to > 23 {
		fmt.Println("invalid move")
		return
	}

	fromCell := field.cells[from]
	toCell := field.cells[to]

	if fromCell.checkers == 0 {
		fmt.Println("invalid move. no checkers on source cell")
		return
	}

	if toCell.checkers > 0 && toCell.color != fromCell.color {
		fmt.Println("invalid move. destination cell is occupied by opponent")
		return
	}

	fmt.Println("Decrementing source cell from ", fromCell.checkers, " to ", fromCell.checkers-1)
	field.cells[from].checkers--
	fmt.Println("Incrementing destination cell from ", toCell.checkers, " to ", toCell.checkers+1)
	field.cells[to].checkers++

	if field.cells[to].color != fromCell.color {
		field.cells[to].color = fromCell.color
	}
}

func (field *GameField) moveCheckerByDice(from int, dice int) {
	if from < 0 || from > 23 {
		fmt.Println("invalid move")
		return
	}

	fromCell := field.cells[from]

	to := 0
	if fromCell.color == White {
		to = from - dice

		if to < 0 {
			to = 24 + to
		}
	} else {
		to = from - dice
	}

	fmt.Println("from:", from, "to:", to)

	field.moveChecker(from, to)
}
